executor/src/witgen/global_constraints.rs

```diff
@@ -268,8 +266,67 @@ fn propagate_constraints<T: FieldElement>(
     (known_constraints, remove)
 }
 
+/// Get simple expression roots for a given expression.
+///
+/// This function recursively evaluates an algebraic expression and attempts
+/// to extract roots from it. The roots are returned as a vector of evaluation
+/// values containing references to algebraic expressions and field elements.
+///
+/// # Parameters
+///
+/// - `expr`: A reference to an `Expression` of type `T`.
+///
+/// # Returns
+///
+/// An `Option` containing a vector of `EvalValue<&AlgebraicReference, T>`
+/// if the roots can be successfully extracted, otherwise `None`.
+pub fn find_algebraic_roots<T: FieldElement>(
+    expr: &Expression<T>,
+) -> Option<Vec<crate::witgen::EvalValue<&AlgebraicReference, T>>> {
+    match expr {
+        Expression::BinaryOperation(AlgebraicBinaryOperation {
+            left,
+            op: AlgebraicBinaryOperator::Mul,
+            right,
+        }) => {
+            let right_root = find_algebraic_roots(right)?;
+            let mut left_roots = find_algebraic_roots(left)?;
+
+            if !left_roots.is_empty() && !right_root.is_empty() {
+                if let ([(id1, Constraint::Assignment(_))], [(id2, Constraint::Assignment(_))]) = (
+                    &left_roots[0].constraints[..],
+                    &right_root[0].constraints[..],
+                ) {
+                    if id1 != id2 {
+                        return None;
+                    }
+                }
+            }
+
+            left_roots.extend(right_root);
+            Some(left_roots)
+        }
+        Expression::Number(_) => Some(vec![]),
+        _ => {
+            let symbolic_ev = SymbolicEvaluator;
+            let root = ExpressionEvaluator::new(symbolic_ev)
+                .evaluate(expr)
+                .ok()
+                .and_then(|l| l.solve().ok())?;
+
+            match root.constraints[..] {
+                [] => Some(vec![]),
```

Comment:
I think in this case, the expression evaluates to zero and this is a very important specila case: While the numbers returned are all roots in that case, the list is not complete (all numbers are roots).

Reply:
So you mean we must return None and don't make any constraint?

Reply:
Yes